### PR TITLE
Fix optional deps in multiple sections

### LINKED ISSUE #5475: Cargo stopped including optional dependencies into build scripts
With `rust-toolchain = nightly-2018-04-30`, it's possible to have an optional dependency specified in `Cargo.toml` as, say,
```
[build-dependencies]
...
serde = { version = "1.0", optional = true }
...
```
and in `build.rs` as, say,
```
#[cfg(feature = "serde")]
extern crate serde;
```
Then `cargo` invokes `rustc` with
```
rustc --crate-name build_script_build build.rs ... --cfg 'feature="serde"' ... --extern serde=...
```

However, with `rust-toolchain = nightly-2018-05-03`, `cargo` invokes `rustc` without the `--extern serde` flag:
```
rustc --crate-name build_script_build build.rs ... --cfg 'feature="serde"' ... 
```
This causes a build error, since the crate can't be found any more.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit fixes an issue where an optional dependency was listed multiple
times in a manifest (multiple sections). This regression was introduced by #5415
and happened because in the resolver we didn't record a `Dependency` as it was
accidentally deduplicated too soon.

The fix here was to ensure that all `Dependency` annotations make their way into
`Resolve` now that we rely on the listed `Dependency` values for correctness.

Closes #5475

### CODE REVIEW

File: src/cargo/core/resolver/context.rs [hunk 1]
```diff
@@ -221,12 +226,16 @@ impl Context {
             ret.push((dep.clone(), base));
         }
 
-        // Any remaining entries in feature_deps are bugs in that the package does not actually
-        // have those dependencies.  We classified them as dependencies in the first place
-        // because there is no such feature, either.
-        if !reqs.deps.is_empty() {
-            let unknown = reqs.deps.keys().map(|s| &s[..]).collect::<Vec<&str>>();
-            let features = unknown.join(", ");
+        // Any entries in `reqs.dep` which wrent' used are bugs in that the
```
Comment (matklad) [inline src/cargo/core/resolver/context.rs]: s/wrent'/weren't/

File: tests/testsuite/build_script.rs [hunk 2]
```diff
@@ -3833,3 +3833,68 @@ fn rename_with_link_search_path() {
...
+        )
+        .file(
+            "bar/src/lib.rs",
+            r#"
+                pub fn bar() -> u32 { 1 }
+            "#,
+        );
+    let p = p.build();
+
+    // assert_that(p.cargo("run"), execs().with_status(0).with_stdout("0\n"));
```
Comment (matklad) [inline tests/testsuite/build_script.rs]: Why is this commented out? We probably should uncomment or delete it?

**Comment by alexcrichton** [comment]
@bors: r=matklad

Indeed, typos!

**Pull request merged**